smart_card_connector_app/src/mock-chrome-api.js

```diff
@@ -0,0 +1,205 @@
+    for (const event of EXTENSION_API_EVENTS) {
+      // Mock chrome.smartCardProviderPrivate.*.addListener()/removeListener().
+      // Cast is to work around the issue that the return type of
+      // createFunctionMock() is too generic (as it depends on the value of the
+      // optional second argument).
+      const mockedAddListener = /** @type {!goog.testing.StrictMock} */ (
+          this.listenersMockControl_.createFunctionMock(
+              `${event}.addListener`));
+      propertyReplacer.setPath(
+          `chrome.smartCardProviderPrivate.${event}.addListener`,
+          mockedAddListener);
+      const mockedRemoveListener = /** @type {!goog.testing.StrictMock} */ (
+          this.listenersMockControl_.createFunctionMock(
+              `${event}.removeListener`));
+      propertyReplacer.setPath(
+          `chrome.smartCardProviderPrivate.${event}.removeListener`,
+          mockedRemoveListener);
+
+      // Check that a listener will be added to the event and the same listener
+      // will be removed.
+      chrome.smartCardProviderPrivate[event].addListener(
+          goog.testing.mockmatchers.isFunction);
+      let listener;
+      mockedAddListener.$once().$does((actual_listener) => {
+        listener = actual_listener;
+        this.listeners_[event] = listener;
+      });
+
+      const argumentMatcher =
+          new goog.testing.mockmatchers.ArgumentMatcher(function(value) {
+            return value === listener;
+          }, 'verifySameListener');
+      chrome.smartCardProviderPrivate[event]
+          .removeListener(argumentMatcher)
+          .$once();
+
+      this.listenersMockControl_.$replayAll();
+    }
+
+    this.functionsMockControl_ = new goog.testing.MockControl();
+    // Mock chrome.smartCardProviderPrivate.report*Result functions.
+    for (const reportFunction of EXTENSION_FUNCTIONS) {
+      // Cast is to work around the issue that the return type of
+      // createFunctionMock() is too generic (as it depends on the value of the
+      // optional second argument).
+      const mockedReport = /** @type {!goog.testing.StrictMock} */ (
+          this.functionsMockControl_.createFunctionMock(`${reportFunction}`));
+      propertyReplacer.setPath(
+          `chrome.smartCardProviderPrivate.${reportFunction}`, mockedReport);
+    }
+
+    // Mock ResultCode enum.
+    for (const resultCode of RESULT_CODES) {
+      propertyReplacer.setPath(
+          `chrome.smartCardProviderPrivate.ResultCode.${resultCode}`,
+          resultCode);
+    }
+  }
+
+  /** @override */
+  disposeInternal() {
+    this.listenersMockControl_.$verifyAll();
+    this.functionsMockControl_.$verifyAll();
+    super.disposeInternal();
+  }
+
+  /**
+   * Dispatches the specified chrome.smartCardProviderPrivate API event with
+   * `args`.
+   * @param {string} eventName
+   * @param  {...*} args
+   */
+  async dispatchEvent(eventName, ...args) {
+    goog.asserts.assertFunction(this.listeners_[eventName]);
```

Comment:
This is stripped out in Release builds. A more reliable alternative is either `GSC.Logging.check` or `assert...()` from `goog.testing.asserts`.

Reply:
thanks